server/base/src/main/java/org/apache/accumulo/server/conf/util/ConfigConverter.java

```diff
@@ -0,0 +1,241 @@
+            zkPropBasePath);
+      }
+    } catch (KeeperException ex) {
+      throw new IllegalStateException(
+          "Failed to convert namespace from ZooKeeper for path: " + zkPathNamespaceBase, ex);
+    } catch (InterruptedException ex) {
+      throw new IllegalStateException(
+          "Interrupted reading namespaces from ZooKeeper for path: " + zkPathNamespaceBase, ex);
+    }
+  }
+
+  public void convertTables() {
+    var zkPathTableBase = zkBasePath + Constants.ZTABLES;
+    try {
+      List<String> tables = zrw.getChildren(zkPathTableBase);
+      for (String table : tables) {
+        String zkPropBasePath = zkPathTableBase + "/" + table + Constants.ZTABLE_CONF;
+        log.info("table:{} base path: {}", table, zkPropBasePath);
+        Map<String,String> props = readLegacyProps(zkPropBasePath);
+        log.info("table props: {} - {}", table, props);
+        writeConverted(PropCacheId.forTable(instanceId, TableId.of(table)), props, zkPropBasePath);
+      }
+    } catch (KeeperException ex) {
+      throw new IllegalStateException(
+          "Failed to convert tables from ZooKeeper for path: " + zkPathTableBase, ex);
+    } catch (InterruptedException ex) {
+      throw new IllegalStateException(
+          "Interrupted reading namespaces from ZooKeeper for path: " + zkPathTableBase, ex);
+    }
+  }
+
+  private void removeLegacyPaths() {
+    for (String path : legacyPaths) {
+      log.debug("delete ZooKeeper path: {}", path);
+      try {
+        zrw.delete(path);
+      } catch (KeeperException ex) {
+        log.warn(
+            "Failed to delete path on property conversion " + path + ", reason" + ex.getMessage());
+      } catch (InterruptedException ex) {
+        Thread.currentThread().interrupt();
+        throw new IllegalStateException(ex);
+      }
+    }
+  }
+
+  private Map<String,String> readLegacyProps(final String path) {
+    requireNonNull(path, "A ZooKeeper path for configuration properties must be supplied");
+    Map<String,String> props = new HashMap<>();
+    try {
+      List<String> children = zrw.getChildren(path);
+      log.info("Looking in: {}, found: {}", path, children);
+      for (String child : children) {
+        if (Property.isValidPropertyKey(child)) {
+          byte[] bytes = zrw.getData(path + "/" + child);
+          props.put(child, new String(bytes, UTF_8));
+          legacyPaths.add(path + "/" + child);
+        } else {
+          log.info("Skipping invalid property: {} in {}", child, path);
+        }
+      }
+    } catch (KeeperException ex) {
+      throw new IllegalStateException("Failed to get children from ZooKeeper for path: " + path,
+          ex);
+    } catch (InterruptedException ex) {
+      throw new IllegalStateException(
+          "Interrupted reading children from ZooKeeper for path: " + path, ex);
+    }
+    return props;
+  }
+
+  private void writeConverted(final PropCacheId propCacheId, final Map<String,String> props,
```

Comment:
I don't think this converter util needs to use PropCacheId. I think it would be simpler if it just used a path directly. A lot of this code might be redundant.

Reply:
I need to look at the conversion / upgrade as a whole before release.